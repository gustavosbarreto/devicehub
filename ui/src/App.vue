<template>
<v-app>
    <v-navigation-drawer v-if="isLoggedIn" :clipped="false" fixed v-model="drawer" :mini-variant="false" enable-resize-watcher app>
        <v-container>
            <div class="text-center">
                <v-icon>mdi-console</v-icon>
                <h2 style="font-family: monospace">ShellHub</h2>
                <span class="overline">beta</span>
            </div>
        </v-container>
        <v-list>
            <v-list-item v-for="item in items" :key="item.title" :to="item.path" two-line>
                <v-list-item-action>
                    <v-icon v-text="item.icon"></v-icon>
                </v-list-item-action>

                <v-list-item-content>
                    <v-list-item-title v-text="item.title"></v-list-item-title>
                </v-list-item-content>
            </v-list-item>
        </v-list>
    </v-navigation-drawer>
    <v-app-bar app elevate-on-scroll color="primary" class="pl-3 pr-4" v-if="isLoggedIn">
        <v-menu transition="scale-transition" origin="top left">
            <template v-slot:activator="{ on }">
                <v-chip v-on="on">
                    <v-icon left>mdi-server</v-icon>
                    My Device Fleet
                    <v-icon right>mdi-chevron-down</v-icon>
                </v-chip>
            </template>
        </v-menu>
        <v-spacer></v-spacer>
        <v-chip>
            <v-icon>help</v-icon>
        </v-chip>
        <v-chip>
            <v-icon>notifications</v-icon>
        </v-chip>
        <v-menu transition="scale-transition" origin="top right">
            <template v-slot:activator="{ on }">
                <v-chip v-on="on">
                    <v-icon left>mdi-account</v-icon>
                    {{ $store.getters["auth/currentUser"] }}
                    <v-icon right>mdi-chevron-down</v-icon>
                </v-chip>
            </template>

            <v-card>
                <v-list-item three-line>
                    <v-list-item-content>
                        <v-list-item-title class="mb-1">Tenant ID</v-list-item-title>
                        <v-list-item-subtitle>
                            <v-chip>
                                <span>
                                    {{ tenant }}
                                </span>
                                <v-icon right v-clipboard="tenant" v-clipboard:success="() => { copySnack = true; }">mdi-content-copy</v-icon>
                            </v-chip>
                        </v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>

                <v-card-actions>
                    <v-btn small text @click="logout()">
                        Logout
                    </v-btn>
                </v-card-actions>
            </v-card>
        </v-menu>
    </v-app-bar>
    <v-content>
        <v-container class="pa-8" fluid>
            <router-view></router-view>
        </v-container>
        <v-snackbar v-model="copySnack" :timeout=3000>Tenant ID copied to clipboard</v-snackbar>
    </v-content>
</v-app>
</template>

<script>
export default {
  name: "App",

  methods: {
    logout() {
      this.$store.dispatch("auth/logout").then(() => {
        this.$router.push("/login");
      });
    }
  },

  computed: {
    tenant() {
      return this.$store.getters["auth/tenant"];
    },

    isLoggedIn() {
      return this.$store.getters["auth/isLoggedIn"];
    }
  },

  data() {
    return {
      drawer: true,
      clipped: false,
      copySnack: false,
      items: [
        {
          icon: "dashboard",
          title: "Dashboard",
          path: "/"
        },
        {
          icon: "devices",
          title: "Devices",
          path: "/devices"
        },
        {
          icon: "history",
          title: "Sessions",
          path: "/sessions"
        }
      ]
    };
  }
};
</script>

<style>
.v-list-active {
  border-left: 4px solid var(--v-primary-base);
}

.text-shadow {
  text-shadow: #000 0 0 6px;
  color: transparent;
}
</style>
