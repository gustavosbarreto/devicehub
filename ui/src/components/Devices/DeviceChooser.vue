<template>
  <v-dialog
    v-if="hasAuthorization"
    v-model="show"
    max-width="900px"
    min-width="45vw"
    data-test="dialog"
  >
    <v-card class="bg-v-theme-surface" data-test="deviceChooser-dialog">
      <v-card-title class="text-headline bg-primary" data-test="title">
        Update account or select three devices
      </v-card-title>

      <v-card-text>
        <p class="ml-2 text-body-2" data-test="subtext">
          You currently have no subscription to the
          <a :href="url()"> premium plan </a> and the free version is limited to
          3 devices. To unlock access to all devices, you can subscribe to the
          <a :href="url()"> premium plan </a>. Case, If you want to continue on
          the free plan, you need to select three devices.
        </p>
      </v-card-text>
      <div v-if="isAllDevices && hasDevice" class="pa-5">
        <v-row>
          <v-col md="12" sm="12">
            <v-text-field
              label="Search by hostname"
              variant="underlined"
              color="primary"
              single-line
              hide-details
              v-model.trim="filter"
              v-on:keyup="searchDevices"
              append-inner-icon="mdi-magnify"
              density="comfortable"
              data-test="search-text"
            />
          </v-col>
        </v-row>
      </div>
      <div class="mt-2">
        <v-tabs align-tabs="center" color="primary" data-test="v-tabs">
          <v-tab
            v-for="item in items"
            :key="item.title"
            :data-test="item.title + '-tab'"
            @click="doAction(item.action)"
          >
            {{ item.title }}
          </v-tab>
        </v-tabs>
      </div>
      <v-card-text class="mb-2 pb-0">
        <DeviceListChooser
          :action="action"
          :isAllDevices="isAllDevices"
          data-test="deviceListChooser-component"
        />
      </v-card-text>

      <v-card-actions>
        <v-spacer />

        <v-btn data-test="close-btn" @click="close()"> Close </v-btn>

        <v-tooltip :disabled="!disableTooltipOrButton" top>
          <template v-slot:activator="{ props }">
            <span v-on="props">
              <v-btn
                v-bind="props"
                :disabled="disableTooltipOrButton"
                data-test="accept-btn"
                @click="accept()"
              >
                Accept
              </v-btn>
            </span>
          </template>

          <span> You can select 3 devices or less. </span>
        </v-tooltip>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script setup lang="ts">
import { ref, computed, onMounted, onUnmounted } from "vue";
import axios, { AxiosError } from "axios";
import { useStore } from "../../store";
import DeviceListChooser from "./DeviceListChooser.vue";
import hasPermision from "../../utils/permission";
import { actions, authorizer } from "../../authorizer";
import { INotificationsSuccess, INotificationsError } from "../../interfaces/INotifications";
import handleError from "../../utils/handleError";
import { IDevice } from "../../interfaces/IDevice";

const store = useStore();
const filter = ref("");
const action = ref("chooser");
const items = ref([
  {
    title: "Suggested Devices",
    action: "suggestedDevices",
  },
  {
    title: "All devices",
    action: "allDevices",
  },
]);

const hostname = ref(window.location.hostname);

const isAllDevices = computed(() => action.value === items.value[1].action);

const show = computed({
  get() {
    return store.getters["devices/getDeviceChooserStatus"];
  },

  set(value) {
    store.dispatch("devices/setDeviceChooserStatus", value);
  },
});

const hasDevice = computed(() => (
  store.getters["stats/stats"].registered_devices > 0
        || store.getters["stats/stats"].pending_devices > 0
        || store.getters["stats/stats"].rejected_devices > 0
));

const url = () => `${window.location.protocol}//${hostname.value}/settings/billing`;

const disableTooltipOrButton = computed(() => (
  (store.getters["devices/getDevicesSelected"].length <= 0
          || store.getters["devices/getDevicesSelected"].length > 3)
        && action.value !== items.value[0].action
));

const hasAuthorization = computed(() => {
  const role = store.getters["auth/role"];
  if (role !== "") {
    return hasPermision(authorizer.role[role], actions.device.chooser);
  }

  return false;
});

const doAction = async (actionParam: string) => {
  action.value = actionParam;
  if (action.value === items.value[0].action) {
    store.dispatch("devices/getDevicesMostUsed");
  } else if (action.value === items.value[1].action) {
    const data = {
      perPage: 5,
      page: 1,
      filter: store.getters["devices/getFilter"],
      status: "accepted",
      sortStatusField: null,
      sortStatusString: "asc",
    };

    try {
      await store.dispatch("devices/setDevicesForUserToChoose", data);
    } catch (error: unknown) {
      if (axios.isAxiosError(error)) {
        const axiosError = error as AxiosError;
        if (axiosError.response?.status === 403) {
          store.dispatch("snackbar/showSnackbarErrorAssociation");
        } else {
          store.dispatch(
            "snackbar/showSnackbarErrorLoading",
            INotificationsError.deviceList,
          );
        }
      }
      handleError(error);
    }
  } else {
    store.dispatch(
      "snackbar/showSnackbarErrorLoading",
      INotificationsError.deviceList,
    );
  }
};

const close = () => {
  store.dispatch("devices/setDeviceChooserStatus", false);
};

const sendDevicesChoice = async (devices: Array<IDevice>) => {
  const choices: Array<string> = [];
  devices.forEach((device) => {
    choices.push(device.uid);
  });

  try {
    await store.dispatch("devices/postDevicesChooser", { choices });
    store.dispatch("snackbar/showSnackbarSuccessAction", INotificationsSuccess.deviceChooser);

    store.dispatch("devices/refresh");

    close();
  } catch (error: unknown) {
    store.dispatch("snackbar/showSnackbarErrorAction", INotificationsError.deviceChooser);
    handleError(error);
  }

  store.dispatch("stats/get");
};

const accept = () => {
  if (action.value === items.value[0].action) {
    sendDevicesChoice(store.getters["devices/getDevicesForUserToChoose"]);
  } else {
    sendDevicesChoice(store.getters["devices/getDevicesSelected"]);
  }
};

const searchDevices = () => {
  let encodedFilter = "";

  if (filter.value) {
    const filterToEncodeBase64 = [
      {
        type: "property",
        params: { name: "name", operator: "contains", value: filter.value },
      },
    ];
    encodedFilter = btoa(JSON.stringify(filterToEncodeBase64));
  }

  try {
    store.dispatch("devices/search", {
      page: store.getters["devices/getPage"],
      perPage: store.getters["devices/getPerPage"],
      filter: encodedFilter,
      status: store.getters["devices/getStatus"],
    });
  } catch {
    store.dispatch("snackbar/showSnackbarErrorDefault");
  }
};

onMounted(async () => {
  action.value = items.value[0].action;
  try {
    await store.dispatch("stats/get");
  } catch (error: unknown) {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      if (axiosError.response?.status === 403) store.dispatch("snackbar/showSnackbarErrorAssociation");
    } else {
      store.dispatch("snackbar/showSnackbarErrorDefault");
    }
    handleError(error);
  }
});

onUnmounted(async () => {
  await store.dispatch("devices/setFilter", null);
});
</script>
