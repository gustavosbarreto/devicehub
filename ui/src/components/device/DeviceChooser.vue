<template>
  <v-dialog
    v-if="hasAuthorization"
    v-model="show"
    persistent
    max-width="900"
  >
    <v-card data-test="deviceChooser-dialog">
      <v-card-title class="headline grey lighten-2 text-center">
        Update account or select three devices
      </v-card-title>

      <v-card-text class="mt-4 pb-1">
        <p>
          You currently have no subscription to the
          <a :href="url()">
            premium plan
          </a> and the free version is limited to 3 devices. To unlock access to all
          devices, you can subscribe to the
          <a :href="url()">
            premium plan
          </a>. Case, If you want to continue on the free plan, you need to select three devices.
        </p>
      </v-card-text>

      <v-app-bar
        flat
        color="transparent"
        class="mt-0"
      >
        <v-tabs
          centered
        >
          <v-tab
            v-for="item in items"
            :key="item.title"
            :data-test="item.title+'-tab'"
            @click="doAction(item.action)"
          >
            {{ item.title }}
          </v-tab>
        </v-tabs>
      </v-app-bar>

      <v-card-text class="mb-2 pb-0">
        <DeviceListChooser
          :action="action"
          data-test="deviceListChooser-component"
        />
      </v-card-text>

      <v-card-actions>
        <v-spacer />

        <v-btn
          text
          data-test="close-btn"
          @click="close()"
        >
          Close
        </v-btn>

        <v-tooltip
          :disabled="!disableTooltipOrButton"
          top
        >
          <template #activator="{ on, attrs }">
            <span v-on="on">
              <v-btn
                text
                v-bind="attrs"
                :disabled="disableTooltipOrButton"
                data-test="accept-btn"
                v-on="on"
                @click="accept()"
              >
                Accept
              </v-btn>
            </span>
          </template>

          <span>
            You can select 3 devices or less.
          </span>
        </v-tooltip>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script>

import DeviceListChooser from '@/components/device/DeviceListChooser';

import hasPermission from '@/components/filter/permission';

export default {
  name: 'DeviceChooserComponent',

  filters: { hasPermission },

  components: {
    DeviceListChooser,
  },

  data() {
    return {
      hostname: window.location.hostname,
      action: 'suggestedDevices',
      dialog: false,
      items: [
        {
          title: 'Suggested Devices',
          action: 'suggestedDevices',
        },
        {
          title: 'All devices',
          action: 'allDevices',
        },
      ],
      permissionAction: 'chooser',
    };
  },

  computed: {
    show: {
      get() {
        return this.$store.getters['devices/getDeviceChooserStatus'];
      },

      set(value) {
        this.$store.dispatch('devices/setDeviceChooserStatus', value);
      },
    },

    disableTooltipOrButton() {
      return (this.$store.getters['devices/getDevicesSelected'].length <= 0
        || this.$store.getters['devices/getDevicesSelected'].length > 3)
        && this.action !== this.items[0].action;
    },

    equalThreeDevices() {
      return this.$store.getters['devices/getDevicesSelected'].length === 3;
    },

    hasAuthorization() {
      const role = this.$store.getters['auth/role'];
      if (role !== '') {
        return hasPermission(
          this.$authorizer.role[role],
          this.$actions.device[this.permissionAction],
        );
      }

      return false;
    },
  },

  watch: {
    show(status) {
      if (status && this.action === this.items[0].action) {
        this.$store.dispatch('devices/getDevicesMostUsed');
      }
    },
  },

  methods: {
    async doAction(action) {
      this.action = action;

      if (action === this.items[0].action) {
        this.$store.dispatch('devices/getDevicesMostUsed');
      } else if (action === this.items[1].action) {
        const data = {
          perPage: 10,
          page: 1,
          filter: this.$store.getters['devices/getFilter'],
          status: 'accepted',
          sortStatusField: null,
          sortStatusString: 'asc',
        };

        try {
          await this.$store.dispatch('devices/setDevicesForUserToChoose', data);
        } catch (error) {
          if (error.response.status === 403) {
            this.$store.dispatch('snackbar/showSnackbarErrorAssociation');
          } else {
            this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.snackbar.deviceList);
          }
        }
      } else {
        this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.snackbar.deviceList);
      }
    },

    accept() {
      if (this.action === this.items[0].action) {
        this.sendDevicesChoice(this.$store.getters['devices/getDevicesForUserToChoose']);
      } else {
        this.sendDevicesChoice(this.$store.getters['devices/getDevicesSelected']);
      }
    },

    async sendDevicesChoice(devices) {
      const choices = [];
      devices.forEach((device) => {
        choices.push(device.uid);
      });

      try {
        await this.$store.dispatch('devices/postDevicesChooser', { choices });
        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.deviceChoose);

        this.$store.dispatch('devices/refresh');

        this.close();
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.deviceChooser);
      }

      this.$store.dispatch('stats/get');
    },

    url() {
      return `${window.location.protocol}//${this.hostname}/settings/billing`;
    },

    close() {
      this.$store.dispatch('devices/setDeviceChooserStatus', false);
    },
  },
};

</script>
