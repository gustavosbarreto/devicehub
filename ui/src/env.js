export default {
  /* eslint-disable no-param-reassign */
  install(Vue) {
    const env = {
      isHosted: window.env.HOSTED === 'true',
    };

    Vue.env = env;
    Vue.prototype.$env = env;
  },
  /* eslint-enable no-param-reassign */
};
