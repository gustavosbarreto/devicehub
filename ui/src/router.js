import Vue from "vue";
import Router from "vue-router";
import Dashboard from "./views/Dashboard.vue";
import store from './store'

Vue.use(Router);

const router = new Router({
    mode: "history",
    base: process.env.BASE_URL,
    routes: [
        {
            path: "/",
            name: "dashboard",
            component: Dashboard
        },
        {
            path: "/devices",
            name: "devices",

            component: () =>
                import(/* webpackChunkName: "devices" */ "./views/Devices.vue")
        },
        {
            path: "/devices/:id",
            name: "featureDevice",
            component: () =>
                import("./views/FeatureDevice.vue") 
        },
        {
            path: "/sessions",
            name: "sessions",
            component: () =>
                import("./views/Sessions.vue")
        },
        {
            path: "/login",
            name: "login",
            component: () =>
                import("./views/Login.vue")
        }
    ]
});

router.beforeEach((to, from, next) => {
    if (to.path !== '/login') {
        if (store.getters['auth/isLoggedIn']) {
          return next()
        }

        return next(`/login?redirect=${to.path}`)
      } else {
        return next()
      }
})

export default router;