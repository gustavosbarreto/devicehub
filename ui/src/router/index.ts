import { RouteRecordRaw, createRouter, createWebHistory } from "vue-router";
import { envVariables } from "../envVariables";
import { store } from "@/store";

const Dashboard = () => import("@/views/Dashboard.vue");
const Devices = () => import("@/views/Devices.vue");
const DeviceList = () => import("@/components/Devices/DeviceList.vue");
const DevicePendingList = () => import("@/components/Devices/DevicePendingList.vue");
const DeviceRejectedList = () => import("@/components/Devices/DeviceRejectedList.vue");
const Containers = () => import("@/views/Containers.vue");
const ContainerList = () => import("@/components/Containers/ContainerList.vue");
const ContainerPendingList = () => import("@/components/Containers/ContainerPendingList.vue");
const ContainerRejectedList = () => import("@/components/Containers/ContainerRejectedList.vue");
const Connectors = () => import("@/views/Connectors.vue");
const detailsConnectors = () => import("@/views/ConnectorDetails.vue");
const DetailsDevice = () => import("@/views/DetailsDevice.vue");
const Sessions = () => import("@/views/Sessions.vue");
const DetailsSessions = () => import("@/views/DetailsSessions.vue");
const FirewallRules = () => import("@/views/FirewallRules.vue");
const PublicKeys = () => import("@/views/PublicKeys.vue");
const Settings = () => import("@/views/Settings.vue");
const SettingProfile = () => import("@/components/Setting/SettingProfile.vue");
const SettingNamespace = () => import("@/components/Setting/SettingNamespace.vue");
const SettingPrivateKeys = () => import("@/components/Setting/SettingPrivateKeys.vue");
const SettingTags = () => import("@/components/Setting/SettingTags.vue");
const SettingBilling = () => import("@/components/Setting/SettingBilling.vue");

export const routes: Array<RouteRecordRaw> = [
  {
    path: "/login",
    name: "login",
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../views/Login.vue"),
  },
  {
    path: "/mfa-login",
    name: "MfaLogin",
    beforeEnter: (to, from, next) => {
      // Check if the user is coming from the login route
      if (from.name === "login") {
        // Allow access to MFA login if the user is coming from the login route
        next();
      } else {
        // Redirect to login if the user is not coming from the login route
        next({ name: "login" });
      }
    },
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../views/MfaLogin.vue"),
  },
  {
    path: "/recover-mfa",
    name: "RecoverMfa",
    beforeEnter: (to, from, next) => {
      if (from.name === "MfaLogin") {
        next();
      } else {
        next({ name: "login" });
      }
    },
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../components/AuthMFA/MfaRecover.vue"),
  },
  {
    path: "/recover-mfa/mail-sucessful",
    name: "RecoverMfaMsg",
    beforeEnter: (to, from, next) => {
      if (from.name === "RecoverMfa") {
        next();
      } else {
        next({ name: "login" });
      }
    },
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../components/AuthMFA/MfaMailRecover.vue"),
  },
  {
    path: "/forgot-pass",
    name: "ForgotPassword",
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../views/ForgotPassword.vue"),
  },
  {
    path: "/validation-account",
    name: "ValidationAccount",
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../views/ValidationAccount.vue"),
  },
  {
    path: "/reset-mfa",
    name: "MfaResetValidation",
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../views/MfaResetValidation.vue"),
  },
  {
    path: "/update-password",
    name: "UpdatePassword",
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../views/UpdatePassword.vue"),
  },
  {
    path: "/sign-up",
    name: "SignUp",
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../views/SignUp.vue"),
  },
  {
    path: "/confirm-account",
    name: "ConfirmAccount",
    meta: {
      layout: "LoginLayout",
      requiresAuth: false,
    },
    component: () => import("../views/ConfirmAccount.vue"),
  },
  {
    path: "/",
    name: "Dashboard",
    component: Dashboard,
  },
  {
    path: "/devices",
    name: "devices",
    component: Devices,
    redirect: {
      name: "listDevices",
    },
    children: [
      {
        path: "",
        name: "listDevices",
        component: DeviceList,
      },
      {
        path: "pending",
        name: "pendingDevices",
        component: DevicePendingList,
      },
      {
        path: "rejected",
        name: "rejectedDevices",
        component: DeviceRejectedList,
      },
    ],
  },
  {
    path: "/containers",
    name: "containers",
    beforeEnter: (to, from, next) => {
      if (!envVariables.isEnterprise && !envVariables.isCloud && envVariables.premiumPaywall) {
        store.commit("users/setShowPaywall", true);
      }
      next();
    },
    component: Containers,
    redirect: {
      name: "listContainers",
    },
    children: [
      {
        path: "",
        name: "listContainers",
        component: ContainerList,
      },
      {
        path: "pending",
        name: "pendingContainers",
        component: ContainerPendingList,
      },
      {
        path: "rejected",
        name: "rejectedContainers",
        component: ContainerRejectedList,
      },
    ],
  },
  {
    path: "/containers/connectors",
    name: "connectors",
    component: Connectors,
    beforeEnter: (to, from, next) => {
      if (!envVariables.isEnterprise && !envVariables.isCloud && envVariables.premiumPaywall) {
        store.commit("users/setShowPaywall", true);
      }
      next();
    },
  },
  {
    path: "/containers/connectors/:id",
    name: "detailsConnectors",
    component: detailsConnectors,
  },
  {
    path: "/device/:id",
    name: "detailsDevice",
    component: DetailsDevice,
  },
  {
    path: "/sessions",
    name: "Sessions",
    component: Sessions,
  },
  {
    path: "/sessions/:id",
    name: "detailsSession",
    component: DetailsSessions,
  },
  {
    path: "/firewall/rules",
    name: "firewalls",
    component: FirewallRules,
    beforeEnter: (to, from, next) => {
      if (!envVariables.isEnterprise && !envVariables.isCloud && envVariables.premiumPaywall) {
        store.commit("users/setShowPaywall", true);
      }
      next();
    },
  },
  {
    path: "/sshkeys/public-keys",
    name: "publicKeys",
    component: PublicKeys,
  },
  {
    path: "/settings",
    name: "settings",
    component: Settings,
    redirect: {
      name: "profileSettings",
    },
    children: [
      {
        path: "profile",
        name: "profileSettings",
        component: SettingProfile,
      },
      {
        path: "namespace-manager",
        name: "namespaceSettings",
        component: SettingNamespace,
      },
      {
        path: "private-keys",
        name: "privateKeysSettings",
        component: SettingPrivateKeys,
      },
      {
        path: "tags",
        name: "tagsSettings",
        component: SettingTags,
      },
      {
        path: "billing",
        name: "billingSettings",
        beforeEnter: (to, from, next) => {
          const enabled = envVariables.billingEnable;
          if (enabled) {
            next();
          } else {
            next("/404");
          }
        },
        component: SettingBilling,
      },
    ],
  },
  {
    path: "/:catchAll(.*)",
    redirect: { name: "NotFound" },
  },
  {
    path: "/404",
    name: "NotFound",
    component: () => import("../views/NotFound.vue"),
  },
];

export const router = createRouter({
  history: createWebHistory(),
  routes,
});

router.beforeEach(async (to, from, next) => {
  const isLoggedIn = store.getters["auth/isLoggedIn"];
  const layout = to.meta.layout || "AppLayout";
  const requiresAuth = to.meta.requiresAuth ?? true;

  await store.dispatch("layout/setLayout", layout);

  // Redirect to the appropriate page based on authentication status
  if (requiresAuth && !isLoggedIn) {
    next({ name: "login" }); // Redirect to login page if authentication is required and user is not logged in
  } else if (to.name === "login" && isLoggedIn) {
    next({ path: "/" }); // Redirect from login page to home if user is already logged in
  } else {
    next(); // Continue with the original navigation
  }
});
