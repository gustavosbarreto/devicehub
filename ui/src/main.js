import Vue from 'vue';

import Snackbar from '@/components/snackbar/Snackbar';

import App from './App';
import router from './router';
import store from './store';
import env from './env';
import './vee-validate';
import vuetify from './plugins/vuetify';

Vue.config.productionTip = false;

Vue.component('Snackbar', Snackbar);

Vue.use(require('vue-moment'));

Vue.use(env);

new Vue({
  vuetify,
  router,
  store,
  render: (h) => h(App),
}).$mount('#app');
