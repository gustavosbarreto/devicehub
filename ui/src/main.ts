import { createApp } from "vue";
import * as Sentry from "@sentry/vue";
import { BrowserTracing } from "@sentry/tracing";
import { envVariables } from "./envVariables";
import vuetify from "./plugins/vuetify";
import { key, store } from "./store";
import { router } from "./router";
import App from "./App.vue";

import { loadFonts } from "./plugins/webfontloader";

import SnackbarComponent from "./components/Snackbar/Snackbar.vue";
import { SnackbarPlugin } from "./plugins/snackbar";

const app = createApp(App);

Sentry.init({
  app,
  dsn: envVariables.sentryDsn || "",
  integrations: [
    new BrowserTracing({
      routingInstrumentation: Sentry.vueRouterInstrumentation(router),
    }),
  ],
  tracesSampleRate: 1.0,
  hooks: ["activate", "create", "destroy", "mount", "update"],
  timeout: 500,
  release: envVariables.version || "latest",
});
Sentry.setTag("project", "shellhub-ui");

loadFonts();

app.use(vuetify);
app.use(router);
app.use(store, key);
app.use(SnackbarPlugin);
app.component("SnackbarComponent", SnackbarComponent);
app.mount("#app");
