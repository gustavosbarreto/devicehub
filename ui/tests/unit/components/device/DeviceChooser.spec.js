import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import DeviceChooser from '@/components/device/DeviceChooser';
import { actions, authorizer } from '../../../../src/authorizer';

describe('DeviceChooser', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);

  let wrapper;

  document.body.setAttribute('data-app', true);

  const role = ['owner', 'administrator'];

  const hasAuthorization = {
    owner: true,
    administrator: false,
  };

  const tests = [
    {
      description: 'Dialog is closes',
      variables: {
        deviceChooserStatus: false,
        devicesSelected: [],
        filter: [],
        devices: [],
        dialog: false,
      },
      data: {
        hostname: 'localhost',
        action: 'suggestedDevices',
        dialog: false,
        items: [
          {
            title: 'Suggested Devices',
            action: 'suggestedDevices',
          },
          {
            title: 'All devices',
            action: 'allDevices',
          },
        ],
        permissionAction: 'chooser',
      },
      computed: {
        disableTooltipOrButton: false,
        equalThreeDevices: false,
      },
      components: {
        'deviceChooserStatus-dialog': false,
      },
      template: {
        'deviceChooserStatus-dialog': false,
        'close-btn': false,
        'accept-btn': false,
      },
    },
  ];

  const storeVuex = (
    deviceChooserStatus,
    devicesSelected,
    filter,
    devices,
    currentrole,
  ) => new Vuex.Store({
    namespaced: true,
    state: {
      deviceChooserStatus,
      devicesSelected,
      filter,
      devices,
      currentrole,
    },
    getters: {
      'devices/getDeviceChooserStatus': (state) => state.deviceChooserStatus,
      'devices/getDevicesSelected': (state) => state.devicesSelected,
      'devices/getFilter': (state) => state.filter,
      'devices/list': (state) => state.devices,
      'auth/role': (state) => state.currentrole,
    },
    actions: {
      'stats/get': () => {},
      'devices/getDevicesMostUsed': () => {},
      'devices/postDevicesChooser': () => {},
      'devices/setDevicesForUserToChoose': () => {},
      'devices/setDeviceChooserStatus': () => {},
      'snackbar/showSnackbarDeviceChooser': () => {},
      'snackbar/showSnackbarErrorAssociation': () => {},
      'snackbar/showSnackbarErrorLoading': () => {},
    },
  });

  tests.forEach((test) => {
    role.forEach((currentrole) => {
      describe(`${test.description} ${currentrole}`, () => {
        beforeEach(() => {
          wrapper = mount(DeviceChooser, {
            store: storeVuex(
              test.variables.deviceChooserStatus,
              test.variables.devicesSelected,
              test.variables.filter,
              test.variables.devices,
              currentrole,
            ),
            localVue,
            stubs: ['fragment'],
            vuetify,
            mocks: {
              $authorizer: authorizer,
              $actions: actions,
            },
          });

          wrapper.setData({ dialog: test.variables.dialog });
        });

        ///////
        // Component Rendering
        //////

        it('Is a Vue instance', () => {
          expect(wrapper).toBeTruthy();
        });
        it('Renders the component', () => {
          expect(wrapper.html()).toMatchSnapshot();
        });

        ///////
        // Data checking
        //////

        it('Compare data with default value', () => {
          Object.keys(test.data).forEach((item) => {
            expect(wrapper.vm[item]).toEqual(test.data[item]);
          });
        });
        it('Process data in the computed', () => {
          Object.keys(test.computed).forEach((item) => {
            expect(wrapper.vm[item]).toEqual(test.computed[item]);
          });
          expect(wrapper.vm.hasAuthorization).toEqual(hasAuthorization[currentrole]);
        });

        //////
        // HTML validation
        //////

        it('Renders the template with components', () => {
          Object.keys(test.components).forEach((item) => {
            expect(wrapper.find(`[data-test="${item}"]`).exists()).toBe(test.components[item]);
          });
        });
        it('Renders the template with data', () => {
          Object.keys(test.template).forEach((item) => {
            expect(wrapper.find(`[data-test="${item}"]`).exists()).toBe(test.template[item]);
          });
        });
      });
    });
  });
});
