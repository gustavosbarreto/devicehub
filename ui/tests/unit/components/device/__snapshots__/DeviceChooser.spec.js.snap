// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceChooser Dialog is closes administrator Renders the component 1`] = `""`;

exports[`DeviceChooser Dialog is closes owner Renders the component 1`] = `
<div role="dialog" class="v-dialog__container">
  <!---->
</div>
`;
